### ISSUE TITLE
sam: Recommend using 0 for mapping quality when record is unmapped

### ISSUE THREAD

**Issue by zaeleus** [samtools/hts-specs#727]
As suggested by @jkbonfield in https://github.com/samtools/hts-specs/issues/715#issuecomment-1508097708:

> as an aside, I note you're also using MAPQ of 255 for "unavailable". Commendable, but my experience is that everyone just uses 0 with unmapped data. I think this is because when FLAG 4 is set the specification states no assumption can be made about MAPQ, so it just feels cleaner to zero it out as all other fields have been.

**Comment by jmarshall** [comment]
I have always thought that the text about “no assumptions for [these fields] in [these circumstances]” is an error that will prevent us from defining meanings for those fields in those currently unspecified circumstances, as discussed in e.g. [this samtools-devel thread](https://sourceforge.net/p/samtools/mailman/samtools-devel/thread/0A3B3627-34E0-494C-825A-807BB41FEAE0%40sanger.ac.uk/#msg24727201).

However over a decade later, I fear the ship may have sailed on changing this policy.

**Comment by d-cameron** [comment]
Although "no assumptions can be made", if one were to be making assumptions, I would interpret a mapq defined over an unmapped read to be the mapping probability of the read being unmapped w.r.t the reference used. That is, reads for which the aligner had candidates and gave up (e.g. bwa when every seed has >500 occurrences in the reference) would be mapq0 (since it has no confidence that it's actually unmapped) and reads which don't match any reference (e.g. contamination, primers), would have non-zero mapq.

**Comment by jkbonfield** [comment]
@d-cameron I agree that could have made sense, but it's too late now and I don't think it's how aligners work so I expect such values wouldn't be sensible calibrated anyway.  Although some proxy based on complexity and length (ie an entropy estimation) could be used as an indicator that it is genuine data not found in this reference (but perhaps is part of this genome, eg from a long insertion).

@jmarshall - also agreed the language about "no meaning" isn't always helpful, but as you say the ship has sailed.  However I think it is reasonable to make recommendations (not requirements) on the values when they are essentially NOPs.  Eg CIGAR has no meaning for unmapped data, but sanity checkers may well gripe about something with CIGAR `151M` and indeed it's been reported as bugs before in aligners, despite it being technically valid SAM.  So a footnote *recommending* the values to use when they have no meaning would be useful and wouldn't be a breaking change.  I'm  unsure with MAPQ though: CIGAR `*` is "unavailable" and is used for unmapped data, so arguably MAPQ `255` is also correct here too, however I'm thinking more for the sake of conformity with existing practice.

**Comment by zaeleus** [comment]
As an extra datapoint, htsjdk 4.1.3 rejects unmapped records with a mapping quality > 0.

```console
$ java -jar picard.jar ViewSam --INPUT in.sam --ALIGNMENT_STATUS All --PF_STATUS All
[...]
[Fri Jan 31 11:58:50 CST 2025] [...] Picard version: Version:3.3.0
Exception in thread "main" htsjdk.samtools.SAMFormatException: Error parsing text SAM file. MAPQ must be zero if RNAME is not specified; File in.sam; Line 2
Line: r1	4	*	0	255	*	*	0	0	NNNN	!!!!
[...]
```

Though this behavior is quite baseless, with respect to the specification.
